Generalize GenericHWControlLoop to all types of RobotHW

RobertWilbrandt opened the pull request:
There is no specific reason why the GenericHWControlLoop should only work with GenericHWInterfaces, as it only needs to call the read() and write() methods of the interface. This change also allows it to be used in a generic runner node for combined_robot_hw, which is useful in #34.

RobertWilbrandt committed:
Generalized GenericHWControlLoop to all types of RobotHW

davetcoleman reviewed (commented):
Good point!

davetcoleman commented:
@AndyZe are you familiar enough to give this a review?

AndyZe commented on include/ros_control_boilerplate/generic_hw_interface.h:
```diff
@@ -80,9 +80,23 @@ class GenericHWInterface : public hardware_interface::RobotHW
   /** \brief Read the state from the robot hardware. */
   virtual void read(ros::Duration &elapsed_time) = 0;
 
+  /** \brief Delegate RobotHW::read() calls to read() */
+  virtual void read(const ros::Time& /*time*/, const ros::Duration& period)
```
Can you please fully doxygenize the comments to explain the arguments better? It's not clear to me what the difference is between elapsed_time (L81), time, and period. Here's my best guess:

```
 /** @brief Delegate RobotHW::read() calls to read() 
  *  @param time Current ROS time. This argument is currently unused, exists as a placeholder
  *  @param period Nominal control loop period
  */
  virtual void read(const ros::Time& /*time*/, const ros::Duration& period)
```

RobertWilbrandt replied:
Sure, i just wanted to make it consistent with the rest of the functions.

AndyZe commented on include/ros_control_boilerplate/generic_hw_interface.h:
```diff
@@ -80,9 +80,23 @@ class GenericHWInterface : public hardware_interface::RobotHW
   /** \brief Write the command to the robot hardware. */
   virtual void write(ros::Duration &elapsed_time) = 0;
 
+  /** \brief Delegate RobotHW::write() calls to write() */
```
Please expand the comment to cover the arguments here too

RobertWilbrandt replied:
Done in revision

AndyZe commented on include/ros_control_boilerplate/generic_hw_interface.h:
```diff
@@ -80,9 +80,23 @@ class GenericHWInterface : public hardware_interface::RobotHW
   virtual void write(ros::Duration &elapsed_time) = 0;
 
+  /** \brief Delegate RobotHW::write() calls to write() */
+  virtual void write(const ros::Time& /*time*/, const ros::Duration& period)
```
why is the first argument necessary? Since I don't see it used in the function.

AndyZe replied:
OK, I understand now. The first argument is there because you overloaded the RobotHW function and wanted to avoid a compiler warning. Well, that's kind of messy but I guess there's nothing to be done about it.

RobertWilbrandt replied:
Hey, thats exactly it. As this project uses c++11 i can add an ```override``` specifier to make this clearer.

AndyZe reviewed (commented):
I think it makes sense to make this more generic. I'll :+1: after you update some comments

davetcoleman commented:
This change breaks API, should we put it in a new ``master`` branch instead, and release it into Noetic only? 

I'm fine just merging it here, though, since this is mostly example code that people use, but don't depend on after forking it. Just playing devils advocate.

RobertWilbrandt committed:
Extend documentation of RobotHW read/write functions

This should illustrate better why these functions are there and have
these signatures

RobertWilbrandt commented:
Hey, just updated the PR with your suggestions.

Regarding the API break: This does not break any existing code, as any call to ```GenericHWControlLoop::GenericHWControlLoop()``` will still work (because ```GenericHWInterface``` inherits from ```RobotHW``` the conversion "just works"). The only other visible change is that ```RobotHW::read(time, period)``` (and write) now actually do something, and i find it pretty unlikely that somebody depended on this function doing absolutely nothing. But feel free to go to another branch if you want to.

AndyZe reviewed (approved):
Looks great to me!
